masu/external/downloader/report_downloader_base.py

```diff
@@ -47,6 +49,61 @@ def __init__(self, download_path=None, **kwargs):
         if 'provider_id' in kwargs:
             self._provider_id = kwargs['provider_id']
 
+    def _get_existing_manifest_db_id(self, assembly_id):
+        """Return a manifest DB object if it exists."""
+        manifest_id = None
+        with ReportManifestDBAccessor() as manifest_accessor:
+            manifest = manifest_accessor.get_manifest(
+                assembly_id,
+                self._provider_id
+            )
+            if manifest:
+                manifest_id = manifest.id
+        return manifest_id
+
+    def check_if_manifest_should_be_downloaded(self, assembly_id):
+        """Check if we should download this manifest.
+
+        We first check if we have a database record of this manifest.
+        That would indicate that we have already downloaded and at least
+        begun processing. We then check the last completed time for
+        a file in this manifest. This second check is to cover the case
+        when we did not complete processing and need to re-downlaod and
+        process the manifest.
+
+        Returns True if the manifest should be downloaded and processed.
+        """
+        manifest_id = None
+        num_processed_files = 0
+        num_total_files = 0
+        today = DateAccessor().today_with_timezone('UTC')
+        last_completed_cutoff = today - datetime.timedelta(hours=1)
+        with ReportManifestDBAccessor() as manifest_accessor:
+            manifest = manifest_accessor.get_manifest(
+                assembly_id,
+                self._provider_id
+            )
+            if manifest:
+                manifest_id = manifest.id
+                num_processed_files = manifest.num_processed_files
+                num_total_files = manifest.num_total_files
+                if num_processed_files < num_total_files:
+                    completed_datetime = manifest_accessor.get_last_report_completed_datetime(
+                        manifest_id
+                    )
+                    if completed_datetime and completed_datetime < last_completed_cutoff:
+                        # It has been more than an hour since we processed a file
+                        # and we didn't finish processing. We should download
+                        # and reprocess.
+                        manifest_accessor.reset_manifest(manifest_id)
```

Comment:
Are you worried about this becoming an infinite loop?

If manifest processing gets to a point where we never finish processing, but always reset and retry, that seems like it could result in an infinite loop.

Reply:
We can certainly boost the `last_completed_cutoff` which would make it take longer before we re-process something. Right now the criteria for this to trigger is that we haven't finished processing any file in the manifest in the last hour. It takes much less than an hour to process a single file in a manifest, so as long as processing is proceeding normally we shouldn't hit this. It really only covers the case where something went wrong (e.g. we turned the worker off when we released new code) and processing never was able to finish.

Reply:
I agree that it's a function of processing time and we're currently well within tolerances.

Because there's a limitation here, I think we ought to monitor for it so that we know when we're approaching that limit. If it's possible to check for this in the code, that's great. But, I think it's reasonable to monitor processing times with Prometheus. (Let's create a new issue to make sure that monitoring gets done, maybe?)